### PR TITLE
At a Glance (smartspace) provider selection support

### PR DESCRIPTION
**Pull request by yasanglass**
## Description

Adds a system that allows users to select the provider of smartspace. The current options are:
1. Lawnchair: The built-in & customizable _At a Glance widget_
2. Google: The _At a Glance_ widget from _the Google app_
3. Google Search: The search bar widget from _the Google app_

The ultimate goal of this feature is not to let users choose between these 3 options as they are already accessible using normal widgets: The goal is to implement a system that can be used later on to introduce new smartspace providers easily. 

## More information
- The two google-powered ones are disabled when the Google app is not installed (or if it is installed but not enabled). 
- If the active smartspace provider is no longer _available_, the active smartspace automatically reverts back to _LawnchairSmartspace_ which is **always** available.
- The two google-powered ones technically keep working even without the Google app but they barely provide any functionality and are pretty much useless, that's why I chose to **not** let users use these smartspace providers without the Google app.
- After some discussion with the core Lawnchair team, I decided to show the name "At a Glance provider" to the user but use "Smartspace Mode" in the actual code as SmartspaceProvider class already exists in the code. Let me know if you think we can name them more clearly.

smartspace-provider

## Type of change

:x: General change (non-breaking change that doesn't fit the below categories like copyediting)
:x: Bug fix (non-breaking change which fixes an issue)
:white_check_mark: New feature (non-breaking change which adds functionality)
:x: Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+         */
+        fun values() = listOf(
+            LawnchairSmartspace,
+            GoogleSmartspace,
+            GoogleSearchSmartspace,
+        )
+    }
+
+    open fun isAvailable(context: Context) = true
+}
```
Comment (Goooler) [inline lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt]:
```suggestion
sealed class SmartspaceMode(
    @StringRes val nameResourceId: Int,
    @LayoutRes val layoutResourceId: Int,
) {
    companion object {
        fun fromString(value: String): SmartspaceMode = when (value) {
            "google" -> GoogleSmartspace
            "google_search" -> GoogleSearchSmartspace
            else -> LawnchairSmartspace
        }

        /**
         * @return The list of all time format options.
         */
        fun values() = listOf(
            LawnchairSmartspace,
            GoogleSmartspace,
            GoogleSearchSmartspace,
        )
    }

    abstract fun isAvailable(context: Context): Boolean
}
```

**Review by SuperDragonXD — APPROVED** [review]

File: lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    override fun toString() = "lawnchair"
+}
+
+object GoogleSearchSmartspace : SmartspaceMode(
+    nameResourceId = R.string.smartspace_mode_google_search,
+    layoutResourceId = R.layout.search_container_workspace,
+) {
+    override fun toString() = "google_search"
+
+    override fun isAvailable(context: Context) =
```
Comment (Goooler) [inline lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt]:
```suggestion
    override fun toString(): String = "google_search"

    override fun isAvailable(context: Context): Boolean =
```

File: lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        context.packageManager.isPackageInstalledAndEnabled("com.google.android.googlequicksearchbox")
+}
+
+object GoogleSmartspace : SmartspaceMode(
+    nameResourceId = R.string.smartspace_mode_google,
+    layoutResourceId = R.layout.smartspace_legacy,
+) {
+    override fun toString() = "google"
+
+    override fun isAvailable(context: Context) =
```
Comment (Goooler) [inline lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt]:
```suggestion
    override fun toString(): String = "google"

    override fun isAvailable(context: Context): Boolean =
```

**Comment by SuperDragonXD** [comment]
Looks good. Although if you long click the Google Search widget, does it show a "preferences" popup that links to this:


? It wasn't shown in the gif.

File: lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    open fun isAvailable(context: Context) = true
+}
+
+
+object LawnchairSmartspace : SmartspaceMode(
+    nameResourceId = R.string.smartspace_mode_lawnchair,
+    layoutResourceId = R.layout.smartspace_container,
+) {
+    override fun toString() = "lawnchair"
```
Comment (Goooler) [inline lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt:39]:
```suggestion
    override fun toString() = "lawnchair"
    override fun isAvailable(context: Context): Boolean = true
```

File: lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt (hunk 1 above)
Comment (Goooler) [reply lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt]: Let's make `SmartspaceMode` as a sealed class.

File: lawnchair/src/app/lawnchair/util/PackageManagerExtensions.kt [hunk 5]
```diff
@@ -21,3 +21,10 @@ fun PackageManager.getPackageVersionCode(packageName: String) =
     } catch (e: PackageManager.NameNotFoundException) {
         -1L
     }
+
+fun PackageManager.isPackageInstalledAndEnabled(packageName: String) =
+    try {
+        getApplicationInfo(packageName, 0).enabled
+    } catch (e: PackageManager.NameNotFoundException) {
+        false
+    }
```
Comment (Goooler) [inline lawnchair/src/app/lawnchair/util/PackageManagerExtensions.kt:30]:
```suggestion
fun PackageManager.isPackageInstalledAndEnabled(packageName: String): Boolean =
    runCatching { getApplicationInfo(packageName, 0).enabled }.getOrDefault(false)
```

**Comment by yasanglass** [comment]
> Looks good. Although if you long click the Google Search widget, does it show a "preferences" popup that links to this?

On my test devices, I can never reach that menu/screen even on the actual widget. What is your device?

File: lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt (hunk 1 above)
Comment (yasanglass) [reply lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt]: Good idea!

File: lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt (hunk 3 above)
Comment (yasanglass) [reply lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt]: I don't see any real benefit in doing this but why not.

Comment (Goooler) [reply lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt]: Prefer declaring return values on public properties.

File: lawnchair/src/app/lawnchair/util/PackageManagerExtensions.kt (hunk 5 above)
Comment (yasanglass) [reply lawnchair/src/app/lawnchair/util/PackageManagerExtensions.kt:30]: I think we should change the style of the functions in this file in another commit outside of this PR.

Comment (Goooler) [reply lawnchair/src/app/lawnchair/util/PackageManagerExtensions.kt:30]: Fine.

**Review by Goooler — APPROVED** [review]

**Comment by SuperDragonXD** [comment]
Well, the screen can be accessible as seen below (Profile Menu > Settings > Search Widget). Not sure what the activity name is though.
https://user-images.githubusercontent.com/70206496/197389485-f387771c-8e84-48bf-9dd5-a0ab16227bca.mp4

**Comment by yasanglass** [comment]
> Well, the screen can be accessible as seen below (Profile Menu > Settings > Search Widget). Not sure what the activity name is though. https://user-images.githubusercontent.com/70206496/197389485-f387771c-8e84-48bf-9dd5-a0ab16227bca.mp4

Just checked and even though we can not just long click on the widget to reach it, it does customize the widget on smartspace.

I will look into adding it.

**Comment by yasanglass** [comment]
After more testing, I discovered that this feature is affected by #2820 since it adds a (configurable) widget to the home screen.

As we cannot just add a fix for #2820 on this pull request & the actual code this pull request adds isn't bugged, I think I should just add this feature as an experimental feature until the widget issue is resolved. Any ideas would be welcome.

**Comment by yasanglass** [comment]
https://github.com/LawnchairLauncher/lawnchair/pull/3049/commits/c65ccdf366968b67e805994ebbaa6e312e4c9240 adds a new experimental preference for smartspace provider selection. This only toggles the visibility of the selection preference in the interface.

**Comment by William150722** [comment]
java.lang.RuntimeException: Unable to stop activity {com.android.launcher}: java.lang.NullPointerException: Attempt to read from field 'com.android.server.appwidget.AppWidgetServiceImpl$ProviderId com.android.server.appwidget.AppWidgetServiceImpl$Provider.id' on a null object reference in method 'android.util.SparseArray com.android.server.appwidget.AppWidgetServiceImpl$Host.getWidgetUids()'
	at android.app.ActivityThread.callActivityOnStop(ActivityThread.java:5422)
	at android.app.ActivityThread.performStopActivityInner(ActivityThread.java:5394)
	at android.app.ActivityThread.handleStopActivity(ActivityThread.java:5459)
	at android.app.servertransaction.StopActivityItem.execute(StopActivityItem.java:43)
	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2374)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:233)
	at android.os.Looper.loop(Looper.java:344)
	at android.app.ActivityThread.main(ActivityThread.java:8248)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:589)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1071)
Caused by: java.lang.NullPointerException: Attempt to read from field 'com.android.server.appwidget.AppWidgetServiceImpl$ProviderId com.android.server.appwidget.AppWidgetServiceImpl$Provider.id' on a null object reference in method 'android.util.SparseArray com.android.server.appwidget.AppWidgetServiceImpl$Host.getWidgetUids()'
	at android.os.Parcel.createExceptionOrNull(Parcel.java:2478)
	at android.os.Parcel.createException(Parcel.java:2456)
	at android.os.Parcel.readException(Parcel.java:2439)
	at android.os.Parcel.readException(Parcel.java:2381)
	at com.android.internal.appwidget.IAppWidgetService$Stub$Proxy.stopListening(IAppWidgetService.java:813)
	at android.appwidget.AppWidgetHost.stopListening(AppWidgetHost.java:258)
	at com.android.launcher3.widget.LauncherAppWidgetHost.stopListening(LauncherAppWidgetHost.java:2)
	at com.android.launcher3.widget.LauncherAppWidgetHost.setShouldListenFlag(LauncherAppWidgetHost.java:7)
	at com.android.launcher3.widget.LauncherAppWidgetHost.setActivityStarted(Unknown Source:1)
	at com.android.launcher3.Launcher.onStop(Launcher.java:7)
	at app.lawnchair.LawnchairLauncher.onStop(LawnchairLauncher.kt:1)
	at android.app.Instrumentation.callActivityOnStop(Instrumentation.java:1510)
	at android.app.Activity.performStop(Activity.java:8371)
	at android.app.ActivityThread.callActivityOnStop(ActivityThread.java:5414)
	... 14 more
Caused by: android.os.RemoteException: Remote stack trace:
	at com.android.server.appwidget.AppWidgetServiceImpl$Host.getWidgetUids(AppWidgetServiceImpl.java:4281)
	at com.android.server.appwidget.AppWidgetServiceImpl.stopListening(AppWidgetServiceImpl.java:814)
	at com.android.internal.appwidget.IAppWidgetService$Stub.onTransact(IAppWidgetService.java:320)
	at android.os.Binder.execTransactInternal(Binder.java:1221)
	at android.os.Binder.execTransact(Binder.java:1163)

@SuperDragonXD similar issue that you mention before? if yes, what to do in order to fix it. Thanks.

**Pull request merged**